fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvPartitionCompensator.java

```diff
@@ -0,0 +1,696 @@
+        for (PartitionKey partitionKey : partitionKeys) {
+            // only support one partition column now.
+            if (partitionKey.getKeys().size() != 1) {
+                return false;
+            }
+            LiteralExpr e = partitionKey.getKeys().get(0);
+            // Only support date/int type
+            if (!(e instanceof DateLiteral || e instanceof IntLiteral)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static List<ScalarOperator> compensatePartitionPredicateForExternalTables(LogicalScanOperator scanOperator) {
+        try {
+            ScanOperatorPredicates scanOperatorPredicates = scanOperator.getScanOperatorPredicates();
+
+            List<ScalarOperator> partitionPredicates = Lists.newArrayList();
+            Table baseTable = scanOperator.getTable();
+            if (baseTable.isUnPartitioned()) {
+                return partitionPredicates;
+            }
+
+            OperatorType operatorType = scanOperator.getOpType();
+            // only used for hive
+            if (operatorType == OperatorType.LOGICAL_HIVE_SCAN) {
+                if (scanOperatorPredicates.getSelectedPartitionIds().size()
+                        == scanOperatorPredicates.getIdToPartitionKey().size()) {
+                    return partitionPredicates;
+                }
+                if (!isSupportedPartitionPredicateForExternalTables(scanOperatorPredicates.getSelectedPartitionKeys())) {
+                    return partitionPredicates;
+                }
+            }
+            List<Range<PartitionKey>> ranges = Lists.newArrayList();
+            for (PartitionKey selectedPartitionKey : scanOperatorPredicates.getSelectedPartitionKeys()) {
+                try {
+                    LiteralExpr expr = PartitionUtil.addOffsetForLiteral(selectedPartitionKey.getKeys().get(0), 1,
+                            PartitionUtil.getDateTimeInterval(baseTable, baseTable.getPartitionColumns().get(0)));
+                    PartitionKey partitionKey = new PartitionKey(ImmutableList.of(expr), selectedPartitionKey.getTypes());
+                    ranges.add(Range.closedOpen(selectedPartitionKey, partitionKey));
+                } catch (AnalysisException e) {
+                    LOG.warn("Compute partition key range failed. ", e);
+                    return partitionPredicates;
+                }
+            }
+
+            List<Range<PartitionKey>> mergedRanges = MvUtils.mergeRanges(ranges);
+            ColumnRefOperator partitionColumnRef = scanOperator.getColumnReference(baseTable.getPartitionColumns().get(0));
+            ScalarOperator partitionPredicate = convertPartitionKeysToPredicate(partitionColumnRef, mergedRanges);
+            if (partitionPredicate != null) {
+                partitionPredicates.add(partitionPredicate);
+            }
+
+            return partitionPredicates;
+        } catch (AnalysisException e) {
```

Comment:
can you narrow down the try-catch scope ?

Reply:
done